Implement Tensor struct

siliconlad opened the pull request.

siliconlad commented:
@CameronMatthew can you have a look and see if this is heading in the right direction.

Very rough around the edges (can you flatten an n-dimensional tensor in your head?), but that can be patched up with refinements to the API and maybe some additional macros.

Also code is a bit of a mess (wth is `iters.rs`) but gotta start somewhere

CameronMatthew commented on src/lib.rs:
```diff
@@ -0,0 +1,924 @@
+    }
+
+    /// For the maths see: https://bit.ly/3KQjPa3
+    fn calculate_index(&self, indices: &[usize]) -> usize {
+        let mut index = 0;
+        for k in 0..self.shape.len() {
+            let stride = self.shape.dims[k+1..].iter().product::<usize>();
+            index += indices[k] * stride;
+        }
+        index
+    }
```
This would potentially be very useful to abstract away for other Tensor types

siliconlad replied:
You mean like sparse tensors?

CameronMatthew replied:
Sparse and static.

CameronMatthew replied:
Actually mostly for static as sparse tensors are a different beast

CameronMatthew replied:
Actually, this could work well - why don't we make this a method for the Shape struct. Since technically all you need is a shape to work out the right index?

siliconlad replied:
We agreed that this should go in a `RawTensor` struct which is responsible for laying out the tensor in memory.

CameronMatthew reviewed (commented):
Definitely along the right lines! This is very nice. One or two places you can make your life easier. Would also like to talk about reusing some of this for Static tensors

Dougal-s commented on src/iter.rs:
```diff
@@ -0,0 +1,82 @@
+pub struct IndexIterator {
+    shape: Vec<usize>,
+    current: Vec<usize>,
+    done: bool,
```
Dynamically allocating memory for every iterator seems like it could be a significant performance problem. I think it would be better to pass the order of the tensor as a template parameter.
```rust
pub struct IndexIterator<const N: usize> {
    shape: [usize; N], // should this be a Shape?
    current: [usize; N],
    done: bool,
}
```
The same thing applies to `Shape` and `Tensor`. Plus, it adds some more type safety.
Granted this does mean that you can't have arrays of different sized tensors, but you can always add extra dimensions of size 1.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,733 @@
+
+impl<T: Num + PartialOrd + Copy> Tensor<T> {
+
+    pub fn new(shape: &Shape, data: &[T]) -> Tensor<T> {
+        assert!(data.len() == shape.size());
+        Tensor {data: data.to_vec(), shape: shape.clone()}
+    }
```
BUT THEY GAVE INCONSISTENT SHAPE INFORMATION AGGGH.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,789 @@
+    pub fn zeros(shape: &Shape) -> Tensor<T> {
+        Tensor::fill(shape, T::zero())
+    }
+
```
Done.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,789 @@
+        Tensor::fill(shape, T::zero())
+    }
+
+    pub fn eye(shape: &Shape) -> Tensor<T> {
+        let mut t = Tensor::zeros(shape);
+        for i in 0..shape.dims[0] {
+            t.set_element(&[i, i], T::one());
+        }
+        t
+    }
```
When your understanding isn't good enough you accidentally get it correct :smile:

siliconlad replied:
I'm looking at this again, and how is it right :laughing:

What exactly do we want this method to return?

siliconlad replied:
Deleted it for now. Will see if I can implement it just for matrices later.

Dougal-s commented on src/lib.rs:
```diff
@@ -0,0 +1,989 @@
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: Tensor<T>) -> Tensor<T> {
+    if self.shape.len() == 1 && rhs.shape.len() == 1 {
```
I think a lot these cases could just be merged? At least, I think that's the case if the code differentiates between row and column vectors.

siliconlad replied:
Yep, definitely needs a rewrite.

siliconlad replied:
They've been split across the different methods in `DynamicVector`, `DynamicMatrix` and `DynamicTensor`.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,789 @@
+
+    // Element-wise operations
+    pub fn pow(&self, power: usize) -> Tensor<T> {
+        let mut t = self.clone();
```
I feel like by default they should create copies. Perhaps a `pow_mut` for in-place ops?

siliconlad replied:
Let's create copies for now. I think this is the better default.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,789 @@
+
+    // Access methods
+    pub fn get_element(&self, indices: &[usize]) -> &T {
+        self.assert_indices(indices).unwrap();
```
BUT THEY GAVE THE WRONG INDICES AAGH.

Dougal-s commented on src/shape.rs:
```diff
@@ -0,0 +1,74 @@
+use std::ops::Index;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Shape {
```
Do we want to treat `N x 1` as different from just `N`?

siliconlad replied:
Good question. What do you think?

siliconlad replied:
I think the consensus was that `N x 1` (and `1 x N`) and `N` are different because they have different "orders". Having said that, I think we will have something which means that `N` can be used in matrix multiplication and other operations as long as it's unambiguous.

For example `N` should be able to multiply with an `N x N` matrix. Whether it's treated as `N x 1` or `1 x N` being based on the order of the multiplication.

The return order is a bit hazy, because numpy would return an order 1 matrix if we use `N`. So maybe we make it consistent with that.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,924 @@
+        self.data[index] = value;
+    }
+
+    // // Reduction operations
+    pub fn sum(&self, axes: Axes) -> Tensor<T> {
+        let all_axes = (0..self.shape.len()).collect::<Vec<_>>();
+        let remaining_axes = all_axes.clone().into_iter().filter(|&i| !axes.contains(&i)).collect::<Vec<_>>();
+        let remaining_dims = remaining_axes.iter().map(|&i| self.shape.dims[i]).collect::<Vec<_>>();
+        let removing_dims = axes.iter().map(|&i| self.shape.dims[i]).collect::<Vec<_>>();
+
+        // We resolve to a scalar value
+        if axes.is_empty() | (remaining_dims.len() == 0) {
+            let sum: T = self.data.iter().fold(T::zero(), |acc, x| acc + *x);
+            return Tensor::new(&Shape::new(vec![1]), &[sum]);
+        }
+
+        // Create new tensor with right shape
+        let new_shape = Shape::new(remaining_dims);
+        let mut t: Tensor<T> = Tensor::zeros(&new_shape);
+
+        for target in IndexIterator::new(&new_shape.dims) {
+            let sum_iter = IndexIterator::new(&removing_dims);
+            for sum_index in sum_iter {
+                let mut indices = target.clone();
+                for (i, &axis) in axes.iter().enumerate() {
+                    indices.insert(axis, sum_index[i]);
+                }
+
+                let value = *t.get_element(&target) + *self.get_element(&indices);
+                t.set_element(&target, value);
+            }
+        }
+
+        t
+    }
```
I'm just surprised I got it working at all.

siliconlad replied:
Wouldn't this require N more memory allocations. One for each intermediate stage?

CameronMatthew replied:
That's true, but is that performance gain worth the sacrifice in complexity/readability at this stage? I'd rather we spent the effort making a sensible abstraction that allowed you to easily do any generic reduction function. Arguably you're optimising for a case that is rare in scientific computing. I make that claim based on very little knowledge of the field in fairness so happy to be wrong here.

siliconlad replied:
Yeah these methods need to be refactored haha.

siliconlad replied:
See #11

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,924 @@
+    } else if self.shape.len() == 2 && rhs.shape.len() == 2 {
+        // Matrix-Matrix multiplication
+        assert!(self.shape[1] == rhs.shape[0], "The number of columns in the first matrix must be equal to the number of rows in the second matrix.");
+        let mut result = Tensor::zeros(&shape![self.shape[0], rhs.shape[1]]);
```
Broadcasting and Tensor slices would be next on the to do list.

CameronMatthew replied:
Speaking of a todo list...

We don't have one.

It will be very easy to get carried away and just implement a bunch of stuff. We should come up with a set of operations to stop at and add more later if there's appetite.

I will start one in an issue.

siliconlad replied:
Yeah I guess that's what we can use this initial implementation for.

CameronMatthew replied:
https://github.com/Rust-Scientific-Computing/feotensor/issues/1#issuecomment-2183283636

Dougal-s commented on src/lib.rs:
```diff
@@ -0,0 +1,989 @@
+pub type Axes = Vec<usize>;
+
+#[derive(Debug, Clone)]
+pub struct Tensor<T: Num> {
```
I think it would be best if we had an `AbstractTensor` trait that `Tensor` implements since we'll have various different types of tensors (e.g. `SparseTensor`, `StaticTensor`).

siliconlad replied:
Yeah I agree. Was thinking of calling it `RawTensor`. That might be semantically slightly different though.

CameronMatthew replied:
That sounds a lot like inheritance - which is not what traits are for. Rather we should have SparseTensors, StaticTensors and DynamicTensors which are composed of reusable parts (like Shape _maybe_).

siliconlad replied:
We probably will want a trait which outlines the basic operations for a generic tensor.

Dougal-s reviewed (commented):
I'm very impressed by how quickly you managed to put this together!

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,789 @@
+        index
+    }
+
+    fn assert_indices(&self, indices: &[usize]) -> Result<(), String> {
+        if indices.len() != self.shape.len() {
+            return Err(format!("Incorrect number of dimensions ({} vs {}).", indices.len(), self.shape.len()));
+        }
+        for (i, &index) in indices.iter().enumerate() {
+            if index >= self.shape.dims[i] {
+                return Err(format!("Index out of bounds for dimension {}", i));
+            }
+        }
+        Ok(())
+    }
```
One advantage of doing it this way is that I can give more helpful errors. Simplifies the `get` and `set` methods as well as I don't need conditional if statements.

siliconlad replied:
Refactored this away.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,924 @@
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: T) -> Tensor<T> {
+        let mut result = Tensor::zeros(&self.shape);
+        for i in 0..self.size() {
+            result.data[i] = self.data[i].clone() * rhs;
+        }
+        result
```
See above.

siliconlad commented:
I have restructured the code to have some of the features that we want. Namely`DynamicStorage` which abstracts away the memory mapping of the tensor. I have also tweaked some of the interfaces for the methods as mentioned above (e.g. returning `Result` in various places).

I'm going to press ahead with some further restructurings:
- [ ] Create dedicated `Vector` and `Matrix` types which support some extra methods
- [ ] Create a distinction between int and float Tensors (to hopefully support `pow)

I'm ignoring the specific implementations of the methods for now because they work (even if they are over-complicated). I will revisit them at the end and tidy them up as well as implement any of the other missing functionality.

siliconlad commented:
Would be good to have another review of the code written so far. I've implemented most of the methods outlined in #1 so far (missing more of the advanced features). Before I add any more, would like to get some feedback on the structure of everything. The way I've done it seems to work, but there's a fair bit of duplication. Perhaps there's a better way.

Largely ignore the code within the methods for now. A lot of it needs to be looked at and refactored. They do get the job done though so :shrug: (I've never had more tests than I have for this repository I don't think ahaha).

@CameronMatthew @Dougal-s

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,789 @@
+    }
+
+    // Element-wise operations
+    pub fn pow(&self, power: usize) -> Tensor<T> {
```
A better `pow` has been implemented, but it is only available if your data type implements the `num::Float` trait, which I think is reasonable for now.

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,924 @@
+use crate::shape::Shape;
+use crate::iter::IndexIterator;
+
+pub type Axes = Vec<usize>;
```
We will still need a vector of axes even if we go with repeated single-axis contractions, unless you want the API to only support one axis at a time (which seems annoying to me). At that point, idk if you want an explicit type like this, or 'just' a `Vec<usize>`.

CameronMatthew commented:
I had a quick look and seems fine. I think we should just merge this asap now.

For the following reasons...
- Something "imperfect" is much better than nothing.
- This PR is getting too large to review properly.
- While this remains on a dev branch, the work you do on it is so broadly scoped that it is impossible for others to contribute without risking massive merge conflicts.
- Merging will allow us to create small, self-contained issues that we can all pick up, work on and review easily.

TL;DR Ship it 🚢.

siliconlad commented:
Sounds good to me.

CameronMatthew reviewed (approved):
LGTM!

siliconlad commented on src/lib.rs:
```diff
@@ -0,0 +1,924 @@
+        }
+
+        t
+    }
```
See #11
